packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -0,0 +1,210 @@
+  /// init message.
+  @visibleForTesting
+  Future<void> connect(Uri connectionUri) async {
+    _channel = WebSocketChannel.connect(
+      connectionUri,
+      protocols: webSocketProtocols,
+    );
+    _subscription = getStreamSubscription(_channel.stream);
+  }
+
+  /// Closes the WebSocket connection.
+  @override
+  void close() {
+    _subscription.cancel();
+    _channel.sink.close();
+  }
+
+  /// Initializes the connection.
+  ///
+  /// Connects to WebSocket, sends connection message and resolves future once
+  /// connection_ack message received from server.
+  Future<void> init() {
+    return _initFuture ??= _init();
+  }
+
+  Future<void> _init() async {
+    final connectionUri =
+        await generateConnectionUri(_config, _authProviderRepo);
+    await connect(connectionUri);
+
+    if (_connectionReady.isCompleted) return;
+    send(WebSocketConnectionInitMessage());
+    return ready;
+  }
+
+  /// Subscribes to the given GraphQL request. Returns the subscription object,
+  /// or throws an [Exception] if there's an error.
+  Stream<GraphQLResponse<T>> subscribe<T>(
+    GraphQLRequest<T> request,
+    void Function()? onEstablished,
+  ) {
+    if (!_connectionReady.isCompleted) {
+      init();
+    }
+
+    final subscriptionId = uuid();
+    return _messageStream
+        .where((msg) => msg.id == subscriptionId)
+        .transform(
+            WebSocketSubscriptionStreamTransformer(request, onEstablished))
+        .asBroadcastStream(
+          onListen: (_) async {
+            // Callout: need to reconsider sending start message onListen.
+            final subscriptionRegistrationMessage =
+                await generateSubscriptionRegistrationMessage(_config,
+                    id: subscriptionId,
+                    authRepo: _authProviderRepo,
+                    request: request);
+
+            send(subscriptionRegistrationMessage);
+          },
+          onCancel: (_) => _cancel(subscriptionId),
+        );
+  }
+
+  /// Cancels a subscription.
+  void _cancel(String subscriptionId) {
+    send(WebSocketStopMessage(id: subscriptionId));
+    // TODO(equartey): if this is the only subscription, close the connection.
+  }
+
+  /// Serializes a message as JSON string and sends over WebSocket channel.
+  @visibleForTesting
+  void send(WebSocketMessage message) {
+    final msgJson = json.encode(message.toJson());
+    _channel.sink.add(msgJson);
+  }
+
+  /// Times out the connection (usually if a keep alive has not been received in time).
+  void _timeout(Duration timeoutDuration) {
+    _rebroadcastController.addError(
+      TimeoutException(
+        'Connection timeout',
+        timeoutDuration,
+      ),
+    );
+  }
+
+  /// Handles incoming data on the WebSocket.
+  ///
+  /// Here, handle connection-wide messages and pass subscription events to
+  /// `_rebroadcastController`.
+  void _onData(WebSocketMessage message) {
```

Comment:
Observation: There are other message types not being handled here. Do we need to handle any of those?

Reply:
Some of the other message types are handled in the transformer, the ones related to individual subscriptions. Are there other types missing?

Reply:
I was thinking of the `complete` type, but looks like you got that handled in that file 👍